In [1]:
from optimize_lib import *
def gradf(x):
    '''
    梯度
    '''

    return np.array([-400*(x[1]-x[0]**2)*x[0]-2*(1-x[0]),200*(x[1]-x[0]**2)],dtype=float)
def f(x):
    '''
    函数
    '''
    return 100*((x[1]-x[0]**2)**2)+(1-x[0])**2

def func(x):
    return 100*(((1+x)-(-1+x)**2)**2)+(2-x)**2

p=0.1   # 设定参数

1. 使用goldstein求第3题$ f(x+\lambda d) $，设定初始$ p=0.1, x=[-1,1]^T, d=[1,1]^T, \lambda=0.1, \alpha=1.5, \beta=0.5 $

In [2]:
x=np.array([-1,1],dtype=float)
d=np.array([1,1],dtype=float)
lamb=goldstein(f,gradf,x,d,0.1,1.5,0.5,p)
print('最终lamb：'+str(lamb))
print('x='+str(x+lamb*d))

lambda的变化情况：[0.1, 0.05, 0.025, 0.0125, 0.00625, 0.003125]
最终lamb：0.003125
x=[-0.996875  1.003125]


2. 使用wolfe-powell求第3题$ f(x+\lambda d) $，设定初始$ x=[-1,1]^T, d=[1,1]^T, \lambda=0.1, \alpha=1.5, \beta=0.5 $

In [3]:
x=np.array([-1,1],dtype=float)
d=np.array([1,1],dtype=float)
lamb=wolf_powell(f,gradf,x,d,0.1,1.5,0.7,p)
print('最终lamb：'+str(lamb))
print('x='+str(x+lamb*d))

lambda的变化情况：[0.1, 0.06999999999999999, 0.048999999999999995, 0.0343, 0.024009999999999997, 0.016806999999999996, 0.011764899999999997, 0.008235429999999997, 0.005764800999999997, 0.004035360699999998, 0.0028247524899999983]
最终lamb：0.0028247524899999983
x=[-0.99717525  1.00282475]


3. 采用goldstein-price，计算$ \sigma=p+0.1 $和$ \sigma=p+0.2 $时的$ \lambda $结果，尝试后发现当其他参数不变$ \sigma>=p+0.2 $时，最终$ \lambda $稳定在0.003125

In [13]:
x=np.array([-1,1],dtype=float)
d=np.array([1,1],dtype=float)

lamb=goldstein_price(f,gradf,x,d,0.1,1.5,0.5,p,p+0.1)
print('最终lamb：'+str(lamb))
print('x='+str(x+lamb*d)+'\n')


lamb=goldstein_price(f,gradf,x,d,0.1,1.5,0.5,p,p+0.2)
print('最终lamb：'+str(lamb))
print('x='+str(x+lamb*d))

lambda的变化情况：[0.1, 0.05, 0.025, 0.0125, 0.00625, 0.003125, 0.004687500000000001, 0.0023437500000000003, 0.0035156250000000005, 0.0052734375, 0.00263671875, 0.003955078125000001]
最终lamb：0.003955078125000001
x=[-0.99604492  1.00395508]

lambda的变化情况：[0.1, 0.05, 0.025, 0.0125, 0.00625, 0.003125]
最终lamb：0.003125
x=[-0.996875  1.003125]


4. 采用黄金分割法

In [5]:
section=[-1,1]
delta=0.001
print('最终lamb：'+str(golden_section(func,section,delta)))

alpha_list:[-1, -0.236, -0.236, -0.236, -0.055695999999999996, -0.055695999999999996, -0.055695999999999996, -0.013108372448000005, -0.013108372448000005, -0.013108372448000005, -0.0030713998625920055, -0.0030713998625920055, 0.0007619608551173596, 0.0007619608551173596, 0.0007619608551173596, 0.0016768140922214835, 0.0016768140922214835]

beta_list:[1, 1, 0.527848, 0.236, 0.236, 0.12457212799999998, 0.055789935999999984, 0.055789935999999984, 0.029470782172863988, 0.013166424895999992, 0.013166424895999992, 0.006963575838217851, 0.006963575838217851, 0.004594558914673463, 0.0031568646171700396, 0.0031568646171700396, 0.002591485316639691]

lambda_list:[-0.236, 0.236, 0.055789935999999984, -0.055695999999999996, 0.055789935999999984, 0.013166424895999992, -0.013108372448000005, 0.013166424895999992, 0.0031568646171700396, -0.0030713998625920055, 0.0031568646171700396, 0.0007619608551173596, 0.0031568646171700396, 0.002226013313867791, 0.0016768140922214835, 0.002226013313867791, 0.0020

1和2返回的结果因设定的步长变化系数α，β而有所差异，由此可推断，不精确的一维搜索的最终步长与设定的步长变化系数有关